cloudinit/config/cc_growpart.py

```diff
@@ -270,36 +270,46 @@ def resize(self, diskdev, partnum, partdev):
                 util.logexc(LOG, "Failed: gpart recover %s", diskdev)
                 raise ResizeFailedException(e) from e
 
-        before = get_size(partdev)
+        before = get_size(partdev, fs)
         try:
             subp.subp(["gpart", "resize", "-i", partnum, diskdev])
         except subp.ProcessExecutionError as e:
             util.logexc(LOG, "Failed: gpart resize -i %s %s", partnum, diskdev)
             raise ResizeFailedException(e) from e
 
-        return (before, get_size(partdev))
+        return (before, get_size(partdev, fs))
+
 
+def get_size(filename, fs) -> Optional[int]:
 
-def get_size(filename) -> Optional[int]:
     fd = None
     try:
         fd = os.open(filename, os.O_RDONLY)
         return os.lseek(fd, 0, os.SEEK_END)
     except FileNotFoundError:
+        if fs == "zfs":
+            return get_zfs_size(filename)
         return None
     finally:
         if fd:
             os.close(fd)
 
 
+def get_zfs_size(dataset) -> Optional[int]:
+    zpool = dataset.split('/')[0]
+    try:
+        size, _ = subp.subp(["zpool", 'get', '-Hpovalue', 'size', zpool])
+    except subp.ProcessExecutionError as e:
+        LOG.debug("Failed: zpool get size %s", zpool)
+        return None
+    return int(size.strip())
+
 def devent2dev(devent):
```

Comment:
might wanna rename this function now that it returns more, maybe `deventinfo`?

Reply:
> might wanna rename this function now that it returns more, maybe deventinfo?

Agreed, how about a snake case `device_entry_info()` or `dev_entry_info()` or something?

Reply:
sounds good, I'll change that tomorrow or the day after.